smart-contracts/wasm-chain-integration/src/utils.rs

```diff
@@ -39,39 +42,60 @@ impl<I> machine::Host<I> for TrapHost {
 /// A host which traps for any function call apart from `report_error` which it
 /// prints to standard out and `get_random` that calls a random number
 /// generator.
-pub struct TestHost<R> {
+pub struct TestHost<'a, R, BackingStore> {
     /// A RNG for randomised testing.
     rng:              Option<R>,
     /// A flag set to `true` if the RNG was used.
     rng_used:         bool,
     /// Debug statements in the order they were emitted.
     pub debug_events: Vec<EmittedDebugStatement>,
+    /// In-memory instance state used for state-related host calls.
+    state:            Option<InstanceState<'a, BackingStore>>,
```

Comment:
Why exactly is this an option?

Can't we always just construct an empty one?

Reply:
I did try making it construct this empty instance state as part of constructing `TestHost`, but I ran into issues with it needing to be self-referential, but I'll give it another attempt.

Reply:
The issue is that `TestHost` needs to own the backing store, the `MutableState` and then the runtime state (`InstanceState`) which holds a reference to both of the before mentioned. It might be possible to get it working, but I think it involves using unsafe.

Reply:
Hmm, I could be wrong, but it seems to me that we never really need to construct the testhost without a state.

We seem to use it in `uninitialized` as a "import export" validator, the same as the the one constructed with ::new. So could we not have these two separate, and then remove the Option here seeing it's only constructed with the actual state ever?

Reply:
I have updated the PR with this last suggestion

Reply:
Looks good.

Reply:
Needs some fixes in benchmarks though.

Reply:
Yes, I'll use `TrapHost` instead of `TestHost` for these.